torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -2236,6 +2253,83 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupGloo::reduce_scatter(
   throw std::runtime_error("ProcessGroupGloo does not support reduce_scatter");
 }
 
+namespace {
+
+class AsyncAlltoallWork : public ProcessGroupGloo::AsyncWork {
+ public:
+  AsyncAlltoallWork(
+      const std::shared_ptr<gloo::Context>& context,
+      at::Tensor& outputTensor,
+      at::Tensor& inputTensor,
+      std::vector<int64_t>& outputCounts,
+      std::vector<int64_t>& inputCounts,
+      uint32_t tag)
+      : context(context),
+        outputTensor(outputTensor),
+        inputTensor(inputTensor),
+        outputCounts(std::move(outputCounts)),
+        inputCounts(std::move(inputCounts)),
+        tag(tag) {}
+
+  std::shared_ptr<gloo::Context> context;
+  at::Tensor outputTensor;
+  at::Tensor inputTensor;
+  std::vector<int64_t> outputCounts;
+  std::vector<int64_t> inputCounts;
+  const uint32_t tag;
+
+  void run() override {
```

Comment:
Seems like the coding style for the other `AsyncWork` classes is to implement all this logic in a helper function like `all_to_all(inputs, outputs)` and then simply call this helper from `run`. Should we use that same style here?

Reply:
Yes, this AsyncAlltoallWork handles both `alltoall` and `alltoallv`. That's why we didn't create a separate function